examples/hosts/iframe-host/src/inner.ts

```diff
@@ -24,26 +28,62 @@ async function getFluidObjectAndRender(container: Container, div: HTMLDivElement
     view.render(div, { display: "block" });
 }
 
-export async function runInner(divId: string) {
-    const div = document.getElementById(divId) as HTMLDivElement;
-
+async function loadContainer(
+    documentId: string,
+    createNew: boolean,
+): Promise<IContainerProxy> {
     const documentServiceFactory = await InnerDocumentServiceFactory.create();
+    const urlResolver = await InnerUrlResolver.create();
+
     const module = { fluidExport: TodoContainer };
     const codeLoader = { load: async () => module };
 
     const loader = new Loader({
-        urlResolver: documentServiceFactory.urlResolver,
+        urlResolver,
         documentServiceFactory,
         codeLoader,
     });
 
-    const url = documentServiceFactory.resolvedUrl.url;
-    const documentId = decodeURIComponent(url.replace("fluid://localhost:3000/tinylicious/", ""));
-    const container = await loader.resolve({ url: documentId });
+    let container: Container;
+
+    if (createNew) {
+        // We're not actually using the code proposal (our code loader always loads the same module regardless of the
+        // proposal), but the Container will only give us a NullRuntime if there's no proposal.  So we'll use a fake
+        // proposal.
+        container = await loader.createDetachedContainer({ package: "no-dynamic-package", config: {} });
+        // Caller is responsible for attaching the created container
+    } else {
+        // Request must be appropriate and parseable by resolver.
+        container = await loader.resolve({ url: documentId });
+        // If we didn't create the container properly, then it won't function correctly.  So we'll throw if we got a
+        // new container here, where we expect this to be loading an existing container.
+        // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+        if (!container.existing) {
+            throw new Error("Attempted to load a non-existing container");
+        }
+    }
+
+    return MakeContainerProxy(container);
+}
 
-    await getFluidObjectAndRender(container, div).catch(() => { });
+async function loadFluidObject(
+    divId: string,
+    containerProxy: IContainerProxy,
+) {
+    const componentDiv = document.getElementById(divId) as HTMLDivElement;
+    await getFluidObjectAndRender(containerProxy, componentDiv).catch(() => { });
     // Handle the code upgrade scenario (which fires contextChanged)
-    container.on("contextChanged", (value) => {
-        getFluidObjectAndRender(container, div).catch(() => { });
+    await containerProxy.on("contextChanged", (value) => {
+        getFluidObjectAndRender(containerProxy, componentDiv).catch(() => { });
     });
 }
+
+export async function runInner(divId: string) {
+    // expose the entrypoints on the iframe window to load a Fluid object
+    (window as any).loadContainer = async (documentId, createNew) => {
```

Comment:
>(window as any) [](start = 4, length = 15)

this shouldn't be necessary

Reply:
our compiler rules don't like us setting arbitrary properties like this :x it will be removed in the followup when it gets moved to use the postmessage api though

---
In reply to: [530567461](https://github.com/microsoft/FluidFramework/pull/4404#discussion_r530567461) [](ancestors = 530567461)